Clean up logic to commit history state from multiple web processes to a shared list in the UI process

charliewolfe opened the pull request:
#### f8b36c9289f3c027fce896b5637ebe6d15505c32
<pre>
Clean up logic to commit history state from multiple web processes to a shared list in the UI process
<a href="https://bugs.webkit.org/show_bug.cgi?id=280876">https://bugs.webkit.org/show_bug.cgi?id=280876</a>
<a href="https://rdar.apple.com/137261520">rdar://137261520</a>

Reviewed by Alex Christensen.

This change is a step towards having a more complete list of back/forward state in the UI process with
site isolation. I added `WebBackForwardListFrameItem` and changed `WebBackForwardListItem` to hold a
reference to it instead of `FrameState`. This will allow the UI process to better manage history state by
frame, as web processes can no longer commit history state for the entire frame tree. More details are
provided below.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::backItem):
(WebCore::BackForwardController::currentItem):
(WebCore::BackForwardController::forwardItem):
(WebCore::BackForwardController::itemAtIndex):
These functions should not be able to request history items for the entire frame tree. Instead, web
processes should only be given history items for the frames they host.

(WebCore::BackForwardController::setChildItem):
This function notifies the UI process when a frame embeds a local child frame.

* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/loader/EmptyClients.cpp:

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::goToItem):
Create and navigate history item trees starting from the root frame, as the main frame may be
out-of-process.

* Source/WebKit/Shared/SessionState.cpp:
(WebKit::FrameState::stateForFrameID const): Deleted.
* Source/WebKit/Shared/SessionState.h:
This function has been replaced with `WebBackForwardListFrameItem::childItemForFrameID`.

* Source/WebKit/Shared/WebBackForwardListItem.cpp:
(WebKit::WebBackForwardListItem::WebBackForwardListItem):
(WebKit::WebBackForwardListItem::~WebBackForwardListItem):
(WebKit::WebBackForwardListItem::itemIsInSameDocument const):
(WebKit::WebBackForwardListItem::itemIsClone):
(WebKit::WebBackForwardListItem::itemID const):
(WebKit::WebBackForwardListItem::setRootFrameState):
(WebKit::WebBackForwardListItem::rootFrameState const):
(WebKit::WebBackForwardListItem::originalURL const):
(WebKit::WebBackForwardListItem::url const):
(WebKit::WebBackForwardListItem::title const):
(WebKit::WebBackForwardListItem::wasCreatedByJSWithoutUserInteraction const):

(WebKit::WebBackForwardListItem::childItemForFrameID const): Deleted.
(WebKit::WebBackForwardListItem::childItemForProcessID const): Deleted.
These functions are replaced by `WebBackForwardListFrameItem::childItemForFrameID`.

* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::setNavigatedFrameID):
(WebKit::WebBackForwardListItem::navigatedFrameID const):
(WebKit::WebBackForwardListItem::mainFrameItem):
(WebKit::WebBackForwardListItem::setIsRemoteFrameNavigation):
(WebKit::WebBackForwardListItem::itemID const): Deleted.
(WebKit::WebBackForwardListItem::setMainFrameState): Deleted.
(WebKit::WebBackForwardListItem::mainFrameState const): Deleted.
(WebKit::WebBackForwardListItem::originalURL const): Deleted.
(WebKit::WebBackForwardListItem::url const): Deleted.
(WebKit::WebBackForwardListItem::title const): Deleted.
(WebKit::WebBackForwardListItem::wasCreatedByJSWithoutUserInteraction const): Deleted.
(WebKit::WebBackForwardListItem::setFrameID): Deleted.
(WebKit::WebBackForwardListItem::frameID const): Deleted.
(WebKit::WebBackForwardListItem::addRootChildFrameItem): Deleted.
(WebKit::WebBackForwardListItem::setMainFrameItem): Deleted.
`m_rootChildFrameItems` is replaced by `m_mainFrameState`.

Also add `m_isRemoteFrameNavigation` to indicate when a back/forward item was created by a navigation
from a process unable to construct history state for the main frame. This change is necessary because
some functions on this object, such as `url()`, have callers that expect this to return the URL for the
main frame.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.cpp:
(WKBackForwardListItemCopyURL):
(WKBackForwardListItemCopyTitle):
(WKBackForwardListItemCopyOriginalURL)
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
(-[WKBackForwardListItem _scrollPosition]):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::backForwardAddItem):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::itemForID):
(WebKit::WebBackForwardList::goToItem):
(WebKit::WebBackForwardList::backForwardListState const):
(WebKit::WebBackForwardList::filteredItemStates const):
(WebKit::WebBackForwardList::addRootChildFrameItem const): Deleted.
* Source/WebKit/UIProcess/WebBackForwardList.h:

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::setPendingChildBackForwardItem):
(WebKit::WebFrameProxy::takePendingChildBackForwardItem):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::hasPendingChildBackForwardItem const):
(WebKit::WebFrameProxy::setHasPendingBackForwardItem): Deleted.
(WebKit::WebFrameProxy::hasPendingBackForwardItem): Deleted.
If a frame expects history state for a child frame to be committed from another process, add the frame
item to `WebFrameProxy`, allowing it to later add the corresponding child item.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::goForward):
(WebKit::WebPageProxy::goBack):

(WebKit::WebPageProxy::goToBackForwardItem):
Update this function to use state from the navigating frame item. The root frame identifier needs to be
sent to the web process, as it is what will have been added to the history item map in the web process.

(WebKit::WebPageProxy::continueNavigationInNewProcess):

(WebKit::WebPageProxy::backForwardAddItem):
(WebKit::WebPageProxy::backForwardAddItemShared):
Change the `WebProcessProxy` parameter to an `IPC::Connection` so we can use it to know when the frame
that has navigated is remote.

(WebKit::WebPageProxy::backForwardSetChildItem):

(WebKit::WebPageProxy::backForwardItemAtIndex):
Return history state for the specified frame rather than sending the entire frame tree.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::updateBackForwardItem):
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::setChildItem):
(WebKit::WebBackForwardListProxy::itemAtIndex):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::itemAtIndex):
* Source/WebKitLegacy/mac/History/WebBackForwardList.mm:
(-[WebBackForwardList itemAtIndex:]):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateFrameWithSiblingsBackForward)):
Add a test for `FIXME: This will not always select the correct child item if there are multiple root
frames in the same process.` which this change resolves.

Canonical link: <a href="https://commits.webkit.org/284750@main">https://commits.webkit.org/284750@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e605a2ef412addad85b51da999e73ef9ee8866a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70319 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49725 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23084 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74408 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21497 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57525 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21337 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55723 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14199 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60623 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36185 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41903 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18058 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19858 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63835 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18402 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76127 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17635 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63442 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14587 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63379 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11418 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5050 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10776 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45527 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/294 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46601 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47878 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46343 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -796,18 +796,18 @@ void Page::setOpenedByDOM()
-    if (RefPtr localMainFrame = dynamicDowncast<LocalFrame>(mainFrame)) {
-        if (localMainFrame->checkedHistory()->shouldStopLoadingForHistoryItem(item))
-            localMainFrame->protectedLoader()->stopAllLoadersAndCheckCompleteness();
+    ASSERT(frame.isRootFrame());
```
I'm not sure this change from main frame to root frame is conceptually right.  There may be many root frames in the same process.  Why don't we instead make a RemoteFrame able to send updates or something like that?

charliewolfe replied:
There may be many root frames, but I don’t think we need to consider root frames that are not ancestors of the frame that is navigating.

achristensen07 commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -877,8 +880,8 @@ void HistoryController::updateBackForwardListClippedAtTarget(bool doClip)
-    Ref mainFrame = frame->mainFrame();
-    Ref topItem = mainFrame->checkedHistory()->createItemTree(page->historyItemClient(), *frame, doClip);
+    Ref rootFrame = frame->rootFrame();
+    Ref topItem = rootFrame->checkedHistory()->createItemTree(page->historyItemClient(), *frame, doClip);
```
I'm not sure that switching from the main frame to the root frame is conceptually right.

charliewolfe replied:
Same as above - I don’t think there is any reason to construct history items for root frames that are not ancestors of the navigating frame.

achristensen07 commented on Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
```diff
@@ -0,0 +1,112 @@
+
+BackForwardItemIdentifier WebBackForwardListFrameItem::identifier() const
+{
+    return *m_frameState->identifier;
```
This should at least have a debug assertion that the identifier is there.
Why not just return a std::optional?

charliewolfe replied:
This is just matching what we already do in `WebBackForwardListItem`. We unconditionally dereference these after [284688@main](https://commits.webkit.org/284688@main).

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9114,19 +9120,29 @@ void WebPageProxy::backForwardAddItemShared(Ref<WebProcessProxy>&& process, Fram
 #endif
 
-    Ref item = WebBackForwardListItem::create(WTFMove(mainFrameState), identifier());
+    if (auto* targetFrame = WebFrameProxy::webFrame(targetFrameID)) {
```
RefPtr instead of auto*

charliewolfe committed:
Clean up logic to commit history state from multiple web processes to a shared list in the UI process
https://bugs.webkit.org/show_bug.cgi?id=280876
rdar://137261520

Reviewed by Alex Christensen.

This change is a step towards having a more complete list of back/forward state in the UI process with
site isolation. I added `WebBackForwardListFrameItem` and changed `WebBackForwardListItem` to hold a
reference to it instead of `FrameState`. This will allow the UI process to better manage history state by
frame, as web processes can no longer commit history state for the entire frame tree. More details are
provided below.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::backItem):
(WebCore::BackForwardController::currentItem):
(WebCore::BackForwardController::forwardItem):
(WebCore::BackForwardController::itemAtIndex):
These functions should not be able to request history items for the entire frame tree. Instead, web
processes should only be given history items for the frames they host.

(WebCore::BackForwardController::setChildItem):
This function notifies the UI process when a frame embeds a local child frame.

* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/loader/EmptyClients.cpp:

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::goToItem):
Create and navigate history item trees starting from the root frame, as the main frame may be
out-of-process.

* Source/WebKit/Shared/SessionState.cpp:
(WebKit::FrameState::stateForFrameID const): Deleted.
* Source/WebKit/Shared/SessionState.h:
This function has been replaced with `WebBackForwardListFrameItem::childItemForFrameID`.

* Source/WebKit/Shared/WebBackForwardListItem.cpp:
(WebKit::WebBackForwardListItem::WebBackForwardListItem):
(WebKit::WebBackForwardListItem::~WebBackForwardListItem):
(WebKit::WebBackForwardListItem::itemIsInSameDocument const):
(WebKit::WebBackForwardListItem::itemIsClone):
(WebKit::WebBackForwardListItem::itemID const):
(WebKit::WebBackForwardListItem::setRootFrameState):
(WebKit::WebBackForwardListItem::rootFrameState const):
(WebKit::WebBackForwardListItem::originalURL const):
(WebKit::WebBackForwardListItem::url const):
(WebKit::WebBackForwardListItem::title const):
(WebKit::WebBackForwardListItem::wasCreatedByJSWithoutUserInteraction const):

(WebKit::WebBackForwardListItem::childItemForFrameID const): Deleted.
(WebKit::WebBackForwardListItem::childItemForProcessID const): Deleted.
These functions are replaced by `WebBackForwardListFrameItem::childItemForFrameID`.

* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::setNavigatedFrameID):
(WebKit::WebBackForwardListItem::navigatedFrameID const):
(WebKit::WebBackForwardListItem::mainFrameItem):
(WebKit::WebBackForwardListItem::setIsRemoteFrameNavigation):
(WebKit::WebBackForwardListItem::itemID const): Deleted.
(WebKit::WebBackForwardListItem::setMainFrameState): Deleted.
(WebKit::WebBackForwardListItem::mainFrameState const): Deleted.
(WebKit::WebBackForwardListItem::originalURL const): Deleted.
(WebKit::WebBackForwardListItem::url const): Deleted.
(WebKit::WebBackForwardListItem::title const): Deleted.
(WebKit::WebBackForwardListItem::wasCreatedByJSWithoutUserInteraction const): Deleted.
(WebKit::WebBackForwardListItem::setFrameID): Deleted.
(WebKit::WebBackForwardListItem::frameID const): Deleted.
(WebKit::WebBackForwardListItem::addRootChildFrameItem): Deleted.
(WebKit::WebBackForwardListItem::setMainFrameItem): Deleted.
`m_rootChildFrameItems` is replaced by `m_mainFrameState`.

Also add `m_isRemoteFrameNavigation` to indicate when a back/forward item was created by a navigation
from a process unable to construct history state for the main frame. This change is necessary because
some functions on this object, such as `url()`, have callers that expect this to return the URL for the
main frame.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.cpp:
(WKBackForwardListItemCopyURL):
(WKBackForwardListItemCopyTitle):
(WKBackForwardListItemCopyOriginalURL)
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
(-[WKBackForwardListItem _scrollPosition]):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::backForwardAddItem):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::itemForID):
(WebKit::WebBackForwardList::goToItem):
(WebKit::WebBackForwardList::backForwardListState const):
(WebKit::WebBackForwardList::filteredItemStates const):
(WebKit::WebBackForwardList::addRootChildFrameItem const): Deleted.
* Source/WebKit/UIProcess/WebBackForwardList.h:

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::setPendingChildBackForwardItem):
(WebKit::WebFrameProxy::takePendingChildBackForwardItem):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::hasPendingChildBackForwardItem const):
(WebKit::WebFrameProxy::setHasPendingBackForwardItem): Deleted.
(WebKit::WebFrameProxy::hasPendingBackForwardItem): Deleted.
If a frame expects history state for a child frame to be committed from another process, add the frame
item to `WebFrameProxy`, allowing it to later add the corresponding child item.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::goForward):
(WebKit::WebPageProxy::goBack):

(WebKit::WebPageProxy::goToBackForwardItem):
Update this function to use state from the navigating frame item. The root frame identifier needs to be
sent to the web process, as it is what will have been added to the history item map in the web process.

(WebKit::WebPageProxy::continueNavigationInNewProcess):

(WebKit::WebPageProxy::backForwardAddItem):
(WebKit::WebPageProxy::backForwardAddItemShared):
Change the `WebProcessProxy` parameter to an `IPC::Connection` so we can use it to know when the frame
that has navigated is remote.

(WebKit::WebPageProxy::backForwardSetChildItem):

(WebKit::WebPageProxy::backForwardItemAtIndex):
Return history state for the specified frame rather than sending the entire frame tree.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::updateBackForwardItem):
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::setChildItem):
(WebKit::WebBackForwardListProxy::itemAtIndex):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::itemAtIndex):
* Source/WebKitLegacy/mac/History/WebBackForwardList.mm:
(-[WebBackForwardList itemAtIndex:]):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateFrameWithSiblingsBackForward)):
Add a test for `FIXME: This will not always select the correct child item if there are multiple root
frames in the same process.` which this change resolves.

Canonical link: https://commits.webkit.org/284750@main
